go/stats/prometheusbackend/collectors.go

```diff
@@ -0,0 +1,204 @@
+	vt   ValueType
+}
+
+// Describe implements Collector.
+func (c *metricsFuncWithMultiLabelsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.desc
+}
+
+// Collect implements Collector.
+func (c *metricsFuncWithMultiLabelsCollector) Collect(ch chan<- prometheus.Metric) {
+	for lvs, val := range c.cfml.Counts() {
+		labelValues := strings.Split(lvs, ".")
+		value := float64(val)
+		ch <- prometheus.MustNewConstMetric(c.desc, toPromValueType(c.vt), value, labelValues...)
+	}
+}
+
+type timingsCollector struct {
+	t    *stats.Timings
+	desc *prometheus.Desc
+}
+
+// Describe implements Collector.
+func (c *timingsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.desc
+}
+
+// Collect implements Collector.
+func (c *timingsCollector) Collect(ch chan<- prometheus.Metric) {
+	for cat, his := range c.t.Histograms() {
+		ch <- prometheus.MustNewConstHistogram(
+			c.desc,
+			uint64(his.Count()),
+			float64(his.Total()),
+			makePromBucket(his.Cutoffs(), his.Buckets()),
+			cat)
+	}
+}
+
+func makePromBucket(cutoffs []int64, buckets []int64) map[float64]uint64 {
+	output := make(map[float64]uint64)
+	last := uint64(0)
+	for i := range cutoffs {
+		key := float64(cutoffs[i]) / 1000000000
+		//TODO(zmagg): int64 => uint64 conversion. error if it overflows?
+		output[key] = uint64(buckets[i]) + last
+		last = output[key]
+	}
+	return output
+}
+
+type multiTimingsCollector struct {
+	mt   *stats.MultiTimings
+	desc *prometheus.Desc
+}
+
+// Describe implements Collector.
+func (c *multiTimingsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.desc
+}
+
+// Collect implements Collector.
+func (c *multiTimingsCollector) Collect(ch chan<- prometheus.Metric) {
+	for cat, his := range c.mt.Timings.Histograms() {
+		labelValues := strings.Split(cat, ".")
+		ch <- prometheus.MustNewConstHistogram(
+			c.desc,
+			uint64(his.Count()),
+			float64(his.Total()),
+			makePromBucket(his.Cutoffs(), his.Buckets()),
+			labelValues...)
+	}
+}
+
+type metricFuncCollector struct {
+	cf   *stats.CounterFunc
+	desc *prometheus.Desc
+	vt   ValueType
+}
+
+// Describe implements Collector.
+func (c *metricFuncCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.desc
+}
+
+// Collect implements Collector.
+func (c *metricFuncCollector) Collect(ch chan<- prometheus.Metric) {
+	ch <- prometheus.MustNewConstMetric(c.desc, toPromValueType(c.vt), float64(c.cf.F()))
+}
+
+type durationFuncCollector struct {
+	df   *stats.DurationFunc
+	desc *prometheus.Desc
+}
+
+// Describe implements Collector.
+func (c *durationFuncCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.desc
+}
+
+// Collect implements Collector.
+func (c *durationFuncCollector) Collect(ch chan<- prometheus.Metric) {
+	// c.df.F() is in nanoseconds. Converting to seconds here for exporting to Prometheus.
+	ch <- prometheus.MustNewConstMetric(c.desc, prometheus.CounterValue, float64(c.df.F()/1000000000.0))
+}
+
+func toPromValueType(vt ValueType) prometheus.ValueType {
```

Comment:
Why do we need this conversion and the `ValueType` enum at all? It seems like holdover from the more generic pullbackend design, and with this current approach we can avoid the conversion by just passing the appropriate prometheus metric type along when we instantiate the collector.